src/prefect/server/utilities/leasing.py

```diff
@@ -66,6 +66,141 @@ async def revoke_lease(self, lease_id: UUID) -> None:
         """
         ...
 
+    async def list_holders_for_limit(self, limit_id: UUID) -> list[dict[str, Any]]:
+        """
+        List all holders for a given limit.
+
+        Args:
+            limit_id: The ID of the limit to list holders for.
+
+        Returns:
+            A list of holder dictionaries with structure {"holder": {...}, "slots": N}
+        """
+        ...
+
+
+# Prefect-specific helper utilities for working with concurrency leases.
+# These functions centralize normalization/fallback logic so API layers stay thin.
+
+from typing import TYPE_CHECKING  # noqa: E402
+
+from prefect.server.database import PrefectDBInterface  # noqa: E402
+from prefect.server.models import (  # noqa: E402
+    concurrency_limits_v2 as cl_v2_models,
+)
+
+if TYPE_CHECKING:  # pragma: no cover
+    from prefect.server.concurrency.lease_storage import ConcurrencyLeaseStorage
+
+
+async def get_active_slots_from_leases(
+    limit_id: UUID, lease_storage: "ConcurrencyLeaseStorage | None" = None
+) -> list[str]:
+    """Extract task_run IDs from active leases for a given limit.
+
+    Handles backends that provide a holder listing helper and those that do not by
+    falling back to scanning active leases.
+    """
+    if lease_storage is None:
+        # Local import to avoid circular dependency during module import
+        from prefect.server.concurrency.lease_storage import (
+            get_concurrency_lease_storage,
+        )
+
+        lease_storage = get_concurrency_lease_storage()
+    # Get holders from storage (now guaranteed by protocol)
+    holders = await lease_storage.list_holders_for_limit(limit_id)
+
+    active_holders: list[str] = []
+    for holder in holders:
+        # Support shapes:
+        # 1) {"type": "task_run", "id": "..."}
+        # 2) {"holder": {"type": "task_run", "id": "..."}, "slots": N}
+        # Since list_holders_for_limit returns dict[str, Any], we know holder is a dict
+        h = holder.get("holder") if "holder" in holder else holder
+
+        if isinstance(h, dict) and h.get("type") == "task_run" and h.get("id"):
+            active_holders.append(str(h["id"]))
+
+    # Deduplicate
+    return list(dict.fromkeys(active_holders))
+
+
+async def find_lease_for_task_run(
```

Comment:
This is used in only one spot, so I'd vote to keep this in `src/prefect/server/api/concurrency_limits.py`.